python/pyarrow/parquet.py

```diff
@@ -729,16 +729,23 @@ def __init__(self, path_or_paths, filesystem=None, schema=None,
 
         self.paths = path_or_paths
 
-        (self.pieces, self.partitions,
-         self.common_metadata_path) = _make_manifest(path_or_paths, self.fs)
+        (self.pieces,
+         self.partitions,
+         self.common_metadata_path,
+         self.metadata_path) = _make_manifest(path_or_paths, self.fs)
 
         if self.common_metadata_path is not None:
             with self.fs.open(self.common_metadata_path) as f:
                 self.common_metadata = ParquetFile(f).metadata
         else:
             self.common_metadata = None
 
-        self.metadata = metadata
+        if metadata is None and self.metadata_path is not None:
+            with self.fs.open(self.metadata_path) as f:
+                self.metadata = ParquetFile(f).metadata
+        else:
+            self.metadata = metadata
```

Comment:
This is potentially confusing since if you pass in `metadata` for schema validation the `metadata_path` could be pointing to a different file than the `metadata` object represents. I think it might be best to have the `metadata` passed into the constructor be strictly used for schema validation and not stored to represent the metadata object of the dataset as it seems like it wouldnt contain the correct row group information of the dataset.

Reply:
When we would go the route and only use the passed-in `metadata` for schema validation, we should probably call the parameter differently (i.e. deprecate the old, add the new one).